File: day2/solve.py
import re

PATTERN = re.compile(r'(\d+)-(\d+) (.): (.+)')

count1 = 0
count2 = 0

with open('input.txt') as f:
    for line in f:
        x, y, letter, password = PATTERN.match(line).groups()
        x = int(x)
        y = int(y)
        if x <= password.count(letter) <= y:
            count1 += 1
        if sum(password[index - 1] == letter for index in [x, y]) == 1:
            count2 += 1

print(count1)
print(count2)

File: day2/solve.rb
PATTERN = /(\d+)-(\d+) (.): (.+)/

lines = File.foreach('input.txt').map do |line|
  x, y, letter, password = PATTERN.match(line).captures
  [x.to_i, y.to_i, letter, password]
end

count1 = lines.count do |min, max, letter, password|
  password.count(letter).between?(min, max)
end

count2 = lines.count do |i, j, letter, password|
  [password[i-1], password[j-1]].count(letter) == 1
end

puts count1
puts count2
